mythril/laser/ethereum/call.py

```diff
@@ -149,28 +147,44 @@ def get_call_data(global_state, memory_start, memory_size, pad=True):
     state = global_state.mstate
     transaction_id = "{}_internalcall".format(global_state.current_transaction.id)
     try:
-        # TODO: This only allows for either fully concrete or fully symbolic calldata.
-        # Improve management of memory and callata to support a mix between both types.
-        calldata_from_mem = state.memory[
-            util.get_concrete_int(memory_start) : util.get_concrete_int(
-                memory_start + memory_size
-            )
-        ]
-        i = 0
-        starting_calldata = []
-        while i < len(calldata_from_mem):
-            elem = calldata_from_mem[i]
-            if isinstance(elem, int):
-                starting_calldata.append(elem)
-                i += 1
-            else:  # BitVec
-                for j in range(0, elem.size(), 8):
-                    starting_calldata.append(Extract(j + 7, j, elem))
+        # Make the assumption that if the memory offset - calldatasize == 0, a delegatecall or similar pattern
+        # is being used, and the entire same calldata should be used.
+        try:
+            if simplify(
+                memory_size - global_state.environment.calldata.calldatasize == 0
+            ):
+                uses_entire_calldata = True
+            else:
+                uses_entire_calldata = False
+        except Z3Exception:
+            uses_entire_calldata = False
+
+        if uses_entire_calldata:
+            logging.debug("Guessing that the entire calldata is being used in a call")
+            call_data_type = CalldataType.SYMBOLIC
+            call_data = global_state.environment.calldata
+        else:
+            # TODO: This only allows for either fully concrete or fully symbolic calldata.
+            # Improve management of memory and callata to support a mix between both types.
+            calldata_from_mem = state.memory[
+                util.get_concrete_int(memory_start) : util.get_concrete_int(
+                    memory_start + memory_size
+                )
+            ]
+            i = 0
```

Comment:
Maybe we could put the logic determining starting_calldata in a separate function, keeps this function from growing to big

Reply:
I think it's small enough (and will only get smaller with fixes taking into account correct memory byte indexing) as is, and I don't have any reason to expect it to get more complicated or grow, or that it will be used elsewhere. Currently that logic is ~11 lines, and it will probably decrease to ~2.

Reply:
Well, IMO this function has a fair bit of nesting and is getting quite lengthy. However, as you say this can be improved with the correct memory implementation